### PR TITLE
54 cap policy exclusions

### PR DESCRIPTION
**Pull request by ssatyapal123**
#  #
Updated AAD Policies 2.1, 2.2, 2.3, 2.4, 2.9, & 2.10 to check for conditional access policy exclusions to users and groups.  Roles will not allow exclusions at this time.

## 🗣 Description ##
If the conditional access policies contain user or group exclusions, the baseline policy will fail.  This change is the first step towards adding the capability for users to define allowable exclusions to these policies. Most agencies will want to include some exclusions  to these policies. Therefore, the automation script should be enhanced to allow for these exclusions.

## 💭 Motivation and context ##
AzureAD uses conditional access policies to evaluate access requests. These policies may include references to existing users, groups, and roles with both an inclusion and exclusion condition. Currently, the conditional access policies assume that all users, groups, and roles should  to meet the policy baselines. 

## 🧪 Testing ##
I tested the changes using the AAD Rego unit tests. Reviewers can run the AAD unit tests for each policy baseline statement. Some issues I ran into included not having all the necessary Rego policies defined in the AADConfig.rego, e.g., 
    "All" in Cap.Conditions.Users.IncludeGroups
    "All" in Cap.Conditions.Users.IncludeRoles
The changes include an update to how the Orchestrator processes the configuration file to include it in the provider export as well as changes to the AAD Rego provider code and unit tests.  Test using variations of the new sample aad-config.yaml test file to ensure proper results for AAD 2.1, 2.2.1, 2.3, 2.4.1, 2.9, & 2.10 with variations of exclusions, both user and group, with and without exemptions in the configuration file.  The most common execution scheme will be running `Invoke-Scuba -ConfigFilePath sample-config-files\aad-config.yaml` while changing the configuration file to elicit the expected results.  Careful when making changes to tenant conditional access policies in testing as it is possible to lock yourself and others out when changing block rules.

## ✅ Pre-approval checklist ##

- [x] This PR has an informative and human-readable title.
- [x] Changes are limited to a single goal - *eschew scope creep!*
- [x] *All* future TODOs are captured in issues, which are referenced
      in code comments.
- [x] All relevant type-of-change labels have been added.
- [x] These code changes follow [cisagov code standards](https://github.com/cisagov/development-guide).
- [x] All relevant repo and/or project documentation has been updated
      to reflect the changes in this PR.
- [x] Tests have been added and/or modified to cover the changes in this PR.
- [x] All new and existing tests pass.

## ✅ Pre-merge checklist ##

- [ ] Revert dependencies to default branches.
- [ ] Finalize version.

## ✅ Post-merge checklist ##

- [ ] Add a tag or create a release.

### CODE REVIEW

File: Rego/AADConfig.rego [hunk 1]
```diff
@@ -117,37 +183,81 @@ tests[{
...
+
+Policies2_2_1[Cap.DisplayName] {
+    Cap := input.conditional_access_policies[_]
+
+    # Match all simple conditions
+    Policy2_2_1ConditionsMatch(Cap)
+
+    # Only match policies with user exclusions if all exempted
+    count(Cap.Conditions.Users.ExcludeGroups) == 0
+    UserExclusionsFullyExempt(Cap, "Policy2_2_1") == true
```
Comment (adhilto) [inline Rego/AADConfig.rego:225]:
```suggestion
    UserExclusionsFullyExempt(Cap, "Policy2_2") == true
```

Needs to be "Policy2_2" to match the key used in the config file.

File: Rego/AADConfig.rego [hunk 2]
```diff
@@ -117,37 +183,81 @@ tests[{
...
+
+Policies2_2_1[Cap.DisplayName] {
+    Cap := input.conditional_access_policies[_]
+
+    # Match all simple conditions
+    Policy2_2_1ConditionsMatch(Cap)
+
+    # Only match policies with user and group exclusions if all exempted
+    UserExclusionsFullyExempt(Cap, "Policy2_2_1") == true
+    GroupExclusionsFullyExempt(Cap, "Policy2_2_1") == true
```
Comment (adhilto) [inline Rego/AADConfig.rego:247]:
```suggestion
    UserExclusionsFullyExempt(Cap, "Policy2_2") == true
    GroupExclusionsFullyExempt(Cap, "Policy2_2") == true
```

Change to match config file key.

File: Rego/AADConfig.rego [hunk 3]
```diff
@@ -208,13 +363,57 @@ tests[{
...
+
+Policies2_4_1[Cap.DisplayName] {
+    Cap := input.conditional_access_policies[_]
+
+    # Match all simple conditions
+    Policy2_4_1ConditionsMatch(Cap)
+
+    # Only match policies with group exclusions if all exempted
+    count(Cap.Conditions.Users.ExcludeUsers) == 0
+    GroupExclusionsFullyExempt(Cap, "Policy2_4_1") == true
```
Comment (adhilto) [inline Rego/AADConfig.rego:405]:
```suggestion
    GroupExclusionsFullyExempt(Cap, "Policy2_4") == true
```

Change to match config key

File: Rego/AADConfig.rego [hunk 4]
```diff
@@ -208,13 +363,57 @@ tests[{
...
+
+Policies2_4_1[Cap.DisplayName] {
+    Cap := input.conditional_access_policies[_]
+
+    # Match all simple conditions
+    Policy2_4_1ConditionsMatch(Cap)
+
+    # Only match policies with user and group exclusions if all exempted
+    UserExclusionsFullyExempt(Cap, "Policy2_4_1") == true
+    GroupExclusionsFullyExempt(Cap, "Policy2_4_1") == true
```
Comment (adhilto) [inline Rego/AADConfig.rego:416]:
```suggestion
    UserExclusionsFullyExempt(Cap, "Policy2_4") == true
    GroupExclusionsFullyExempt(Cap, "Policy2_4") == true
```

Change to match config key

File: Testing/Unit/Rego/AAD/AADConfig2_02_test.rego [hunk 5]
```diff
@@ -5,16 +5,746 @@ import future.keywords
...
+            "ServicePlanId": "31a0d5b2-13d0-494f-8e42-1e9c550a1b24"
+        },
+        {
+            "ServicePlanName": "AAD_PREMIUM_P2",
+            "ServicePlanId": "c7d91867-e1ce-4402-8d4f-22188b44b6c2"
+        }
+        ],
+        "scuba_config": {
+            "Aad": {
+                 "Policy2_2_1": {
```
Comment (adhilto) [inline Testing/Unit/Rego/AAD/AADConfig2_02_test.rego:248]:
```suggestion
                 "Policy2_2": {
```

File: Testing/Unit/Rego/AAD/AADConfig2_02_test.rego [hunk 6]
```diff
@@ -25,31 +755,94 @@ test_Conditions_Correct if {
...
```
Comment (adhilto) [inline Testing/Unit/Rego/AAD/AADConfig2_02_test.rego:813]:
```suggestion
                 "Policy2_2": {
```

**Review by adhilto — CHANGES_REQUESTED** [review]
- A few comments are minor readability improvements. These are marked with "optional." Do what you will with these.
- The bulk of my comments deal with some inconsistency in how the exclusion keys are named. In the example config file, they are named `Policy2_x`. However, inside the rego, they are `Policy2_1`, **`Policy2_2_1`**, `Policy2_3`, **`Policy2_4_1`**, `Policy2_9`, and `Policy2_10`. This is one change I would like to see addressed. In my commit suggestions, I recommended 2.2 and 2.4 be renamed to match the `Policy2_x` format used in the example config file, though I suppose it could go the other way.

File: Rego/AADConfig.rego [hunk 7]
```diff
@@ -82,27 +82,93 @@ ReportDetailsBooleanLicenseWarning(Status) = Description if {
...
 
+################
+# User/Group Exclusion support functions
+################
+default UserExclusionsFullyExempt(_, _) := false
+UserExclusionsFullyExempt(Policy, PolicyID) := true if {
+    ExemptedUsers := input.scuba_config.Aad[PolicyID].CapExclusions.Users
+    ExcludedUsers := { x | x := Policy.Conditions.Users.ExcludeUsers[_] }
+    AllowedExcludedUsers := { y | y := ExemptedUsers[_] }
+    AllowedExcludedUsers
```
[the remark this answers is not in the document]
Comment (adhilto) [reply Rego/AADConfig.rego:98]: Changing all to `Policy2_x_y` would address my concern, thanks

**Review by adhilto — APPROVED** [review]
Thanks for address my feedback, looks great

**Comment by nanda-katikaneni** [comment]
merge tests passed (note that there are additional policy failures on a given tenant but these new fails are consistent with implemented cap exclusions) -> merging into the main.

**Pull request merged**